### ISSUE TITLE
MIGRATED: The WebView has no support for Basic Authentication (or custom headers). (#19200)

### ISSUE THREAD

**Issue by jamonholmgren** [react-native-webview/react-native-webview#18]
Migrated from https://github.com/facebook/react-native/issues/19200.

@mausworks says:

---

 I have reviewed the documentation
  I have searched existing issues
  I am using the latest React Native version
This is kind of a duplicate of issue #8693, an almost two year issue outlining the same problem. I have left a comment there regarding this. But just to make it clear:

Given how common hybrid apps are, and how common Basic Authentication is, I honestly think this deserves more attention.

That issue also does not conform to your template (likely because of its age), and it has no steps outlined for reproduction, so I figured I'd open a new issue for this; concerning Basic Authentication only.

There is currently no way to use Basic Authentication in the WebView!

It is possible to pass custom headers in the first request via source.headers; but this only applies to the very first request. Meaning: external resources, such as script files, style-sheets, images, videos, etc.— will not be loaded; because they fail authorization. This renders source.headers useless in many cases.

Personally, though; think this is "as expected". The property is named "source" after all. Instead, I think it should be possible to provide the WebView with "default headers": Headers that will be included with every request. Or via an onRequest-callback, which would be a more "catch-all"-solution.

Environment
OS: Windows 10
Node: 8.9.4
Yarn: Not Found
npm: 6.0.0
Watchman: Not Found
Xcode: N/A
Android Studio: Version 3.1.0.0 AI-173.4697961

Steps to Reproduce
An important feature is missing! However, one could argue that the source property of the WebView should forward headers with each consecutive request: Because it is currently the only mechanism which allows for this. Just to make it super clear: I don't think this should be the case, but in order to better explain the issue, I will outline it as if it should be.

I have created a proof-of-concept repository, which can be used to reproduce the bug.

First: Create a server which requires Basic Authentication for all resources. Then have it serve HTML with scripts and/or styles referenced via <link href="..."> or <script src="..."> tags.

Then: Create an application with a WebView in React Native, have it request the web server with Basic Authentication, like so:

const source = {
  uri: "http://<your-host-with-basic-auth>/index.html",
  headers: {
    Authorization: "Basic Zm9vOmJhcg==" // base64 of 'foo:bar'
  }
};

export default class App extends React.Component {
  render() {
    return <WebView source={source} style={{ flex: 1 }} />;
  }
}
Expected Behavior
I expect the CSS and JS files to be loaded, i.e. for the headers provided in source to be included with subsequent requests to these resources.

If my example were to be used, then this is what i expect in my web view:

![](https://user-images.githubusercontent.com/8259221/39810201-76f7eb3e-5384-11e8-853c-dc6fcf9f27a6.png)

Actual Behavior
No CSS or JS is loaded, i.e. the headers are not included in the subsequent requests.

This is the actual output:

![](https://user-images.githubusercontent.com/8259221/39810694-3e54e8ca-5386-11e8-9f86-a836169bfb47.png)

---

**Comment by santiago** [comment]
You may just do http://username:password@host/what/ever

**Comment by jamonholmgren** [comment]
@santiago Thank you! Is that the recommended workaround for this issue? If so, we'll probably want to document that somewhere.

**Comment by Titozzz** [comment]
I know there is restriction for this on older ios versions. So we should double check @santiago

**Comment by jamonholmgren** [comment]
React Native only supports back to iOS 9, so prior to that we won't have to support either.

> Supported operating systems are >= Android 4.1 (API 16) and >= iOS 9.0.

**Comment by Titozzz** [comment]
I'll make some test today and let you guys know.

**Comment by mausworks** [comment]
Hello! Thanks for migrating the issue. 👍 

As most of you have pointed out, `username:password@host.tld` only works on some iOS devices, I've not gotten it to work on any Android device.

Also, while I've argued from a point of view that "basic authentication does not work", but a more fundamental problem is that *custom headers cannot be sent*. The best solution (in my opinion), would be to add support for them, that is: Headers which are sent with *every* request, not just the first one.

This is a bigger technical challenge though, and I fully understand if you want to defer it, or not do it at all.

**Comment by Titozzz** [comment]
For me auth works on android using the URL

I agree that we definitely need to handle custom headers properly.
If you thank to participate in that feature feel free to :)

I'll keep this issue open til I make so tests

**Comment by schnika** [comment]
@mausworks I had no issue to get the PROTOCOL://USERNAME:PASSOWRD@host.tld working on andoid and ios.
This is our implementation.

```
if (process.env.NODE_ENV === 'development') {
  baseUri = 'https://xx:xxxx@asdf.de';
}

...
return(
        <WebView
          useWebKit
          source={{
            uri: baseUri,
          }}
        />
)
```

**Comment by Titozzz** [comment]
Yeah. Have you tried iOS 9 and 10 ?

**Comment by mausworks** [comment]
@schnika @Titozzz this is not a viable solution, it's a work-around.

**Comment by jackyshek** [comment]
My case is that the header does not implement to the request at all. I have saw the RNCWebViewManager.java which the header should be implemented to the webview. Is there any workaround on that 
`<WebView
      source={{uri:"http://localhost:8080/index.jsp", headers:{"test":"header"}, method:"GET"}}
/>`

**Comment by nar213** [comment]
@mausworks have you started working on a PR for either of your two design proposals?

**Comment by jamonholmgren** [comment]
We would love help on fixing this issue as both @Titozzz and I have very busy day jobs. Our role here is mainly to review, merge, and deploy pull requests, and respond to issues as we can. The community's role is to push the webview forward. Thanks to everyone who has contributed already!

**Comment by mausworks** [comment]
@nar213

> @mausworks have you started working on a PR for either of your two design proposals?

God no &mdash; I'm just here to complain, and I'll continue to complain until somebody else solves _my problem_. 🙃

Jokes aside, I have never worked with "real native app development", and frankly &mdash; I don't want to. I'm a web / back-end guy, which occasionally (reluctantly) have to do some native app development.

AFAIC, the problem is solved (or worked around). We figured it was easier to just implement "real" token-based authentication, and have a log-in in the web app, than it was to make the WebView pass custom headers. It was further down the road map anyways, so we just moved it up.

**Comment by nar213** [comment]
@mausworks your solution seems like a reasonable workaround. For developers creating a react-native application that is closely linked to a website the ability to pass custom headers is fairly important for maintaining consistent experience between native and non-native screens. 

PR #181 looks like it is a good starting point for solving the problem

**Comment by jerrychong25** [comment]
Same issue.

I am looking for a Basic Authentication for WebView solution too.

**Comment by MakhouT** [comment]
Also need this. Is there a workaround?

**Comment by jamonholmgren** [comment]
Bump to keep this open.

*Status: still waiting on a PR from the community to provide basic auth and custom headers.*

Please note that the workaround posted (`username:password@example.com`) works in some cases, but not all cases.

References: 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#Access_using_credentials_in_the_URL

![image](https://user-images.githubusercontent.com/1479215/65184576-a7290900-da1a-11e9-94e0-8959c46c0f3d.png)

https://bugs.chromium.org/p/chromium/issues/detail?id=82250#c7

**Comment by Gevie** [comment]
In the mean time if you wish to use basic authentication within the WebView, you may try the following workaround. 

Track the URL in a component state and update it with the requested URL to cause the WebView to be re-rendered sending our headers with each request. 

```
class TestCase extends Component {

    state = {
        url: 'http://yourhosthere.tld'
    };

    render() {
        return(
            <WebView
                source={{
                    uri: this.state.url,
                    headers: { Authorization: 'value' }
                }}
                onShouldStartLoadWithRequest={this.handlePageLoad}
            />
        )
    }

    handlePageLoad = navState => {
        this.setState({ url: navState.url });
        return false;
    }
}
```

I grab the URL from `onShouldStartLoadWithRequest` then return false to ensure the page never loads within the WebView. You can handle this inline rather than separately but in my own case I do much than just setState here.

**Comment by Kebabpizza** [comment]
Why was this closed?

**Comment by jste7102** [comment]
Agree why was this closed? I am also facing this issue currently.

**Comment by rickhanlonii** [comment]
Re-opening, I don't think this was intentionally closed

**Comment by keremcubuk** [comment]
I have the same issue. It's not working on iOS 13.  If I used webview without token it works, but I have to use token. For ex:
```javascript
<WebView
   source={{
     uri: 'https://www.mydomain.com/xxxx',
     headers: { Authorization: requestHeader.headers.Authorization },
   }}
/>
```

Any idea?

**Comment by jamonholmgren** [comment]
Are you forming your Authorization header properly, @keremcubuk ? It has to be in the format:

`Basic <Base64 encoded "username:password" string>`

**Comment by jamonholmgren** [comment]
Additionally, if you need to send it for subsequent page loads, use the technique here: 

https://github.com/react-native-community/react-native-webview/blob/82af9559ef17692e4d1f653e47154071cbd9ea33/docs/Guide.md#working-with-custom-headers-sessions-and-cookies

**Comment by keremcubuk** [comment]
> Are you forming your Authorization header properly, @keremcubuk ? It has to be in the format:
> 
> `Basic <Base64 encoded "username:password" string>`

Hi @jamonholmgren,

I'm using using bearer jwt token. And I've also tried `custom-headers` but it didn't work

My format is:
```
Authorization: Bearer <myJWTtoken>
```

**Comment by jamonholmgren** [comment]
Hm, not really sure, to be honest. Try debugging server-side by printing out the headers you're receiving?

**Comment by rickhanlonii** [comment]
@jamonholmgren any idea why this keeps getting closed by the bot?

**Comment by JakeStoeffler** [comment]
Can this be reopened?  I might have spare bandwidth to work on it soon.  The workarounds suggested above allow me to include custom headers with subsequent page loads in the webview, but that's only one of the problems here.  As the author of this issue put it, "external resources, such as script files, style-sheets, images, videos, etc.— will not be loaded; because they fail authorization"... in other words, any resource loaded by the webview that's not considered a "page load" (aka the url of the webview did not change) will not have the custom headers sent with its request, even if the above workarounds are used.  Again this seems like a common use case that shouldn't require a hack to achieve.

**Comment by mausworks** [comment]
@JakeStoeffler there are probably a quite few people eagerly waiting for this to be solved, I'm sure (see also #4)! For me, we just went with cookie-based authenticated (we had control over the back-end), so it's a non-issue on my part. 

But if you hare the spare bandwidth, then I'm sure a lot of people would be happy to have this resolved &mdash; me included, as I've been getting notifications for this since May 2018, and it would be really cool to see this resolved.

**Comment by iwashi1t** [comment]
Since it seemed difficult to implement custom headers(especially in iOS WKWebView), I created a PR (#1467) to support only basic authentication for the moment.

**Comment by lfaz** [comment]
Is there still a solution for this? I have tried this https://github.com/react-native-webview/react-native-webview/blob/master/docs/Guide.md#Setting-Custom-Headers and it does not work!

**Comment by redbar0n** [comment]
@lfaz The basic authentication support (by @iwashi1t) is now merged in, if it helps: https://github.com/react-native-webview/react-native-webview/pull/1467

**Comment by AMoktar** [comment]
is there any solutions ?

**Comment by Sreelakshmi192** [comment]
Still facing the same issue even with the  latest version

**Comment by abdoerrahiem** [comment]
any updates?

**Comment by nicefella** [comment]
any updates on this?

**Comment by Gevie** [comment]
This post has been open for a long time and I've been following it closely as it was an issue I experienced over 1.5 years ago. This may not be helpful but maybe it is... We got around this by sharing session / cookies between the web application and the webview app (and a mix of JWT/API calls for authentication as well as some other stuff), it took a bit but it was possible.

I'd have preferred a real solution however.

**Comment by khari998** [comment]
> 

Can you show your workaround by any chance? I am looking for a way to achieve this myself

**Comment by khari998** [comment]
> This post has been open for a long time and I've been following it closely as it was an issue I experienced over 1.5 years ago. This may not be helpful but maybe it is... We got around this by sharing session / cookies between the web application and the webview app (and a mix of JWT/API calls for authentication as well as some other stuff), it took a bit but it was possible.
> 
> I'd have preferred a real solution however.

Can you show your workaround by any chance? I am looking for a way to achieve this myself

**Comment by doodlemoonch** [comment]
@khari998 [this workaround](https://github.com/react-native-webview/react-native-webview/blob/82af9559ef17692e4d1f653e47154071cbd9ea33/docs/Guide.md#working-with-custom-headers-sessions-and-cookies) worked for me, ([example component](https://gist.github.com/doodlemoonch/32fc65c778cb5386c4f482857778202a))
